chapters/high_level_shader_language_comparison.adoc

```diff
@@ -0,0 +1,972 @@
+// Copyright 2023 Sascha Willems
+// SPDX-License-Identifier: CC-BY-4.0
+
+ifndef::chapters[:chapters:]
+ifndef::images[:images: images/]
+
+[[shader-decoder-ring]]
+= Vulkan High Level Shader Language Comparison
+:toc:
+
+While Vulkan itself consumes shaders in a binary format called xref:{chapters}what_is_spirv.adoc[SPIR-V], shaders are usually written in a high level language. This section provides a mapping between shader functionality for the most common ones used with Vulkan: GLSL and HLSL. This is mostly aimed at people wanting to migrate from one high level shader language to another. It's meant as a starting point and not as a complete porting guide to one language from another
+
+// @todo: maybe also add SPIR-V
+
+[TIP]
+====
+For more details on using HLSL with Vulkan, visit xref:{chapters}hlsl.adoc[this chapter].
+====
+
+[NOTE]
+====
+The following listings are by no means complete, and mappings for newer extensions may be missing. Also note that concepts do not always map 1:1 between GLSL and HLSL. E.g. there are no semantic in GLSL while some newer GLSL functionality may not (yet) be available in HLSL.
+====
+
+== Extensions
+
+In GLSL extensions need to be explicitly enabled using the `#extension` directive. This is **not** necessary in HLSL. The compiler will implicitly select suitable SPIR-V extensions based on the shader. If required one can use `-fspv-extension` arguments to explicitly select extensions.
+
+== Data types
+
+[NOTE]
+====
+Types work similar in GLSL and HLSL. But where GLSL e.g. has explicit vector or matrix types, HLSL uses basic types. On the other hand HLSL offers advanced type features like C++ templates. This paragraph contains a basic summary with some examples to show type differences between the two languages.
+====
+
+[options="header"]
+|====
+| *GLSL* | *HLSL* | *Example*
+| vec__n__ | float__n__ | vec4 -> float4
+| ivec__n__ | int__n__ | ivec3 --> int3
+| mat__nxm__ or shorthand mat__n__ | float__nxm__ | mat4 -> float4x4
+|====
+
+* link:https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-data-types[HLSL data types (Microsoft)]
```

Comment:
Matrix types in GLSL and HLSL are different in the sense that in the shader code, HLSL is row-major, while GLSL is column-major. This affects construction and access. mat3x4 in GLSL is 3 columns of vec4, while in HLSL float3x4 is 3 rows of float4.